### DISCUSSION TITLE
How to configure shadow for large mesh-like peer-to-peer networks?

### DISCUSSION THREAD

**Discussion by zyansheep** [shadow/shadow#2810] [Q&A]
Hi, I've been using shadow in development of my peer-to-peer protocol and its been really helpful so far :)

My protocol is designed in such a way as to prioritize nodes to mostly connect to their nearby nodes (latency-wise) as to make it hard for the network to be publicly enumerable. Instead of having a publicly-maintained list of nodes (like TOR), I am using [network coordinates](https://orbi.uliege.be/bitstream/2268/136727/1/phdthesis.pdf) to discover and route to nodes in different locations.

 I am now ready to start testing larger networks (100-1000 nodes), and I am curious if anyone here knows how I might generate shadow.yaml files, and how I can find .gml files that model the topology of the internet (with preferably residential nodes as endpoints) that I can use to simulate a mesh-like overlay networks on top of an backbone router topology.

Any advice would be greatly appreciated! :D

**Comment by stevenengler** [comment]
> Hi, I've been using shadow in development of my peer-to-peer protocol and its been really helpful so far :)

Glad to hear it, that's what Shadow's designed for :)

> I am now ready to start testing larger networks (100-1000 nodes), and I am curious if anyone here knows how I might generate shadow.yaml files

We have some documentation on dynamically generating shadow configuration files at: https://shadow.github.io/docs/guide/shadow_config_complex.html#dynamic-generation

This is the same technique used in Tor network simulations by [tornettools](https://github.com/shadow/tornettools), although the tornettools generation has much more Tor-specific code.

> and how I can find .gml files that model the topology of the internet (with preferably residential nodes as endpoints) that I can use to simulate a mesh-like overlay networks on top of an backbone router topology.

The network used by tornettools is the [`atlas_v201801.shadow_v2.gml.xz`](https://github.com/tmodel-ccs2018/tmodel-ccs2018.github.io/blob/master/data/shadow/network/atlas_v201801.shadow_v2.gml.xz) network, which was generated as part of ["Privacy-Preserving Dynamic Learning of Tor Network Traffic"](https://www.robgjansen.com/publications/tmodel-ccs2018.pdf) in 2018 using the [RIPE Atlas](https://atlas.ripe.net/) network. If you search the paper for "atlas", you'll find some details about how this was generated. I don't know much more about this, but if you have any questions @robgjansen will know best.

**Reply by robgjansen** [reply]
I am invoked :)

Yes, Section 6.1.1 in [the referenced paper](https://www.robgjansen.com/publications/tmodel-ccs2018.pdf) provides the best explanation of how we formed the atlas graph. It is meant to be a general, high-level model of Internet latency and bandwidths. Vertices in the graph represent cities around the world and model up/downstream bandwidth, and edges model latencies between every pair of vertices. Each vertex represents a city, but really the ripe atlas probe from which we were able to collect measurements. Sometimes those are run by individuals, sometimes by organizations such as universities.

The graph is a simplified representation of the Internet, but is now ~5 years old. You could of course construct your own graph, but the existing one should be enough to get started.

**Reply by zyansheep** [reply]
Hmm, that should probably be good enough for now... not sure if I want to do the whole 1000 nodes just yet though, Do you know any good tools for editing graphml files?

**Reply by stevenengler** [reply]
We usually use the networkx python library. It does use a lot of memory for large graphs.
